pywbem/_logging.py

```diff
@@ -0,0 +1,402 @@
+
+          log_dest (:term:`string`):
+            String defining the destination for this log. It must be one of the
+            destinations defined in LOG_DESTINATIONS
+
+          log_filename (:term:`string`):
+            Filename to use as logging file if the log destination is `file`.
+            Ignored if log destination is not `file`. The default is
+            DEFAULT_LOG_FILENAME defined in config.py
+
+          log_detail_level (:term:`string`):
+            String defining the level of detail for log output. This is
+            optional.
+            Not all of the loggers use this attribute.  Thus, the ops and http
+            loggers used a defined detail level and always set logging to that
+            level.
+
+          log_level (:term:`string`):
+            String defining the log level. This is optional.  Not all of the
+            loggers use this attribute.  Thus, the ops and http loggers used a
+            defined level and always set logging to that level.
+
+        Return:
+            None
+
+        Exceptions:
+            ValueError - Input cannot be mapped to a defined logger or
+            component.
+            No named logger is created.
+        """
+        if log_dest not in LOG_DESTINATIONS:
+            raise ValueError('Invalid log destination %s. Not in list %s' %
+                             (log_dest, LOG_DESTINATIONS))
+        if log_component not in PYWBEM_LOG_COMPONENTS:
+            raise ValueError('Invalid log component %s. Not in list %s' %
+                             (log_component, PYWBEM_LOG_COMPONENTS))
+        if not log_level:
+            log_level = DEFAULT_LOG_LEVEL
+        if log_level not in LOG_LEVELS:
+            raise ValueError('Invalid log level %s. Not in list %s' %
+                             (log_level, LOG_LEVELS))
+
+        if not log_detail_level:
+            log_detail_level = DEFAULT_LOG_DETAIL_LEVEL
+        if log_detail_level not in LOG_DETAIL_LEVELS:
+            raise ValueError('Invalid log detail %s. Not in list %s' %
+                             (log_detail_level, LOG_DETAIL_LEVELS))
+
+        # If destinations or components is all, recurse for all destinations
+        if log_dest == 'all':
+            for dest in LOG_DESTINATIONS:
+                cls.create_logger(log_component, dest,
+                                  log_filename=log_filename,
+                                  log_detail_level=log_detail_level,
+                                  log_level=log_level)
+        elif log_component == 'all':
+            for comp in PYWBEM_LOG_COMPONENTS:
+                if comp != 'all':
+                    cls.create_logger(comp, log_dest=log_dest,
+                                      log_filename=log_filename,
+                                      log_detail_level=log_detail_level,
+                                      log_level=log_level)
+
+        # Otherwise process results of any recursive calls above
+        else:
```

Comment:
Te checks between lines 245 and 262 can really be moved to here. In their current position, they are needlessly performed one extra time in case of components = all, plus it is more logical to have then here, and maybe some of them can be merged with some of the other checks here.